stl/inc/istream

```diff
@@ -617,14 +617,15 @@ public:
             return -1;
         }
 
-        bool _Sync_failed = true; // sync fails if an exception is thrown
-        _TRY_IO_BEGIN
-        _Sync_failed = _Rdbuf->pubsync() == -1;
-        _CATCH_IO_END
-        if (_Sync_failed) {
+        _TRY_BEGIN
+        if (_Rdbuf->pubsync() == -1) {
             _Myios::setstate(ios_base::badbit);

```

Comment:
I am concerned about calls to `_Myios::setstate()` within try-catch blocks. As far as I can tell (from a quick search), the existing code carefully performs try-catch separately from `_Myios::setstate()`. As I understand it, `setstate()` can throw (if configured to do so via `exceptions()`) so this could lead to a freshly-thrown exception being immediately caught. It might then be rethrown, but the situation seems avoidable.

This similarly applies below. I am not 100% sure that this is an issue (perhaps 90% sure).

Reply:
Hm that explains a lot of the horrible things that are done all over the code.

That said I am not 100% sure about the possibility of it actually throwing. the code in question comes from `<xiosbase>`
```cpp

    void __CLR_OR_THIS_CALL clear(iostate _State, bool _Reraise) { // set state, possibly reraise exception
        _State &= _Statmask;
        _Mystate             = _State;
        const auto _Filtered = _State & _Except;
        if (_Filtered) {
            if (_Reraise) {
                _RERAISE;
            }

            const char* _Msg;
            if (_Filtered & ios_base::badbit) {
                _Msg = "ios_base::badbit set";
            } else if (_Filtered & ios_base::failbit) {
                _Msg = "ios_base::failbit set";
            } else {
                _Msg = "ios_base::eofbit set";
            }

            _THROW(failure(_Msg));
        }
    }
```

So the only possibility that setstate could throw is if the `_Except` bit is set. However, the only place where the `_Execpt` bit is set, is `exceptions(iostate _Newexcept)` (as you correctly stated). Now `exceptions(iostate _Newexcept)` by itself calls `clear()`  immediately afterwards and throws there.

So from what I can tell a call to `setstate` can only throw if `exceptions(iostate _Newexcept)` has previously been called. 

So lets assume that for whatever reason `_Except` is set.

1. If we call `_Myios::setstate(ios_base::badbit)` inside the `_TRY` block we would end up with `_THROW(failure(_Msg))`, as `_Reraise` is `false`. We will then end up in the `_CATCH_ALL` block and call `_Myios::setstate(ios_base::badbit, true)`. This time `_Reraise` is true and we end up with `_RERAISE` which throws the exception emited by `_THROW(failure(_Msg))`.

2. If we only set a temporary state `_State = ios_base::badbit` and then call `_Myios::setstate(_State)` outside of the catch block, we will directly end up in `_THROW(failure(_Msg))`

So in both cases the state will be set to `ios_base::badbit` and we will leave `sync()` via an exception thrown by `_THROW(failure(_Msg))`

Is that analysis correct?

Reply:
I think so, but I just remembered that there's one difference if we `_RERAISE` (i.e. `throw;`) - the debugger won't show the original exception throw site.

I am nervous about this new pattern as it's inconsistent with existing practice, but I can't point to concrete badness and I'm not an iostreams expert, so I won't object.

Reply:
Lets just go with the safe thing and create the `_State` temporary and use that.

It is not the worst thing we have done for conformance

Reply:
I am not really happy, because now we always must call `_Myios::setstate`

Also I believe that if in between there was an exception we would already be in the situation that we essentially just `throw`

Reply:
I've approved, but I'm concerned about this "invisible invariant". Could we document this somewhere? We really should start thinking about how to document the implementation. As the library itself gets larger - not to mention the number of contributors - I don't think we can rely on Stephan to scale.

Reply:
I believe it would be a good followup to expand the test for `exceptions` and then see whether we really need all that complexity